pyspark_datasources/googlesheets.py

```diff
@@ -0,0 +1,160 @@
+        url_query = {"tqx": "out:csv"}
+        if self.sheet_id:
+            url_query["gid"] = self.sheet_id
+        if query:
+            url_query["tq"] = query
+        return f"{path}?{urlencode(url_query)}"
+
+
+class GoogleSheetsDataSource(DataSource):
+    """
+    A DataSource for reading table from public Google Sheets.
+
+    Name: `googlesheets`
+
+    Schema: By default, all columns are treated as strings and the header row defines the column names.
+
+    Examples
+    --------
+    Register the data source.
+
+    >>> from pyspark_datasources import GoogleSheetsDataSource
+    >>> spark.dataSource.register(GoogleSheetsDataSource)
+
+    Load data from a public Google Sheets document using url.
+
+    >>> url = "https://docs.google.com/spreadsheets/d/10pD8oRN3RTBJq976RKPWHuxYy0Qa_JOoGFpsaS0Lop0/edit?gid=0#gid=0"
+    >>> spark.read.format("googlesheets").options(url=url)
+    +-------+----------+-----------+--------------------+
+    |country|  latitude|  longitude|                name|
+    +-------+----------+-----------+--------------------+
+    |     AD| 42.546245|   1.601554|             Andorra|
+    |    ...|       ...|        ...|                 ...|
+    +-------+----------+-----------+--------------------+
+
+    Load data using `spreadsheet_id` and optional `sheet_id`.
+
+    >>> spark.read.format("googlesheets").options(spreadsheet_id="10pD8oRN3RTBJq976RKPWHuxYy0Qa_JOoGFpsaS0Lop0", sheet_id="0")
+    +-------+----------+-----------+--------------------+
+    |country|  latitude|  longitude|                name|
+    +-------+----------+-----------+--------------------+
+    |     AD| 42.546245|   1.601554|             Andorra|
+    |    ...|       ...|        ...|                 ...|
+    +-------+----------+-----------+--------------------+
+
+    Specify custom schema.
+
+    >>> spark.read.format("googlesheets").options(url=url).schema("id string, lat double, long double, name string")
+    +---+----------+-----------+--------------------+
+    | id|       lat|       long|                name|
+    +---+----------+-----------+--------------------+
+    | AD| 42.546245|   1.601554|             Andorra|
+    |...|       ...|        ...|                 ...|
+    +---+----------+-----------+--------------------+
+    """
+
+    @classmethod
+    def name(self):
+        return "googlesheets"
+
+    def __init__(self, options: Dict[str, str]):
+        if "url" in options:
+            self.sheet = Sheet.from_url(options["url"])
+        elif "spreadsheet_id" in options:
+            self.sheet = Sheet(options["spreadsheet_id"], options.get("sheet_id"))
+        else:
+            raise ValueError(
+                "You must specify a URL or spreadsheet_id in `.options()`."
+            )
+
+    def schema(self) -> StructType:
+        import pandas as pd
+
+        # Read schema from the first row of the sheet
+        df = pd.read_csv(self.sheet.get_query_url("select * limit 0"))
```

Comment:
What if the spreadsheet is empty?

Reply:
This returns nothing if the spreadsheet is empty. So this line would fail with `pandas.errors.EmptyDataError`. I'll add a test for this case.